GCRCatalogs/helpers/tract_catalogs.py

```diff
@@ -0,0 +1,65 @@
+"""
+Helper functions for tract catalogs.
+"""
+
+import numpy as np
+
+from GCR import GCRQuery
+
+__all__ = ["tract_filter", "sample_filter"]
+
+
+def tract_filter(tracts, tract_high=None):
+    """
+    Returns a GCRQuery object to be used in the `native_filters` argument of get_quantities(),
+    to select only the tracts in *tracts* (a list of integers).
+
+    If *tracts* is a single integer, select only that tract.
+    If *tracts* and *tract_high* are both given as single integers, select [tracts, tract_high]
+    (inclusive on both ends!).
+    """
+    if isinstance(tracts, int):
+        if tract_high is None:
+            return GCRQuery('tract == {}'.format(tracts))
+        elif isinstance(tract_high, int):
+            return GCRQuery('tract >= {}'.format(tracts), 'tract <= {}'.format(tract_high))
+        raise ValueError("When `tracts` is an integer, `tract_high` must be an integer or None.")
+
+    tracts = np.unique(np.asarray(tracts, dtype=np.int))
+    if not tracts.size:
+        raise ValueError("Must select at least one tract.")
+
+    def _tract_selector(tract, tracts_to_select=tracts):
+        return np.in1d(tract, tracts_to_select, assume_unique=True)
+
+    return GCRQuery((_tract_selector, "tract"))
+
+
+def sample_filter(frac, random_state=None):
+    """
+    Returns a GCRQuery object to be used in the `filters` argument of get_quantities()
+    to randomly sample the object catalog by a given fraction (*frac*).
+
+    Optionally, provide *random_state* (int or np.random.RandomState) to fix random state.
+    """
+    # pylint: disable=no-member
+
+    frac = float(frac)
+    if frac > 1 or frac < 0:
+        raise ValueError("`frac` must be a float number in [0, 1].")
+    if frac == 1:
+        return GCRQuery()
+    if frac == 0:
+        return GCRQuery((lambda a: np.zeros_like(a, dtype=np.bool), "tract"))
+
+    if not isinstance(random_state, np.random.RandomState):
+        random_state = np.random.RandomState(random_state)
+    seed = random_state.randint(65536)
+
+    def _sampler(tract, frac=frac, seed=seed):
+        size = len(tract)
```

Comment:
I don't have in my mind clearly how this will be called.

But if a user were directly calling this then I would be concerned about this assumption about the type of `tract`.
1. If `tract` is an int, then `len` will raise `TypeError`
2. If `tract` is a string, then `len` will give the length of the string
3. If `tract` is an array, then `len` will give the length of the array
Sorry, you know all this.  I'm half just writing this out for myself.  But I was a bit concerned about the type safety here.

Reply:
Because the next line seems to be implying that if `size!=0` then we return `....rand(size)` but if `size==0` then we return `np.zeros`.  But we only get `size==0` if passed an empty list (empty collection).  When would that happen?

Reply:
Thanks for raising this, @wmwv. `_sampler` is not a user facing function, and it's only used in `GCRQuery` (see the return line). Recall that for `GCRQuery`, one can supply a tuple that consists of a callable + columns to feed into that callable. Here, `_sampler` is that callable, and `tract` is the column of tract IDs that will be sent to `_sampler`. The `_sampler` must return a boolean mask array that have the same length as the input array. 

Hence, `tract`, which I now rename to `tract_arr`, will always be a numpy array of tract IDs. For a given tract, it'll be the same value. In fact, for the purpose of sampling, I could choose the `objectId` column as well. I am choosing `tract` because this function can them be applied to any catalog that has `tract` (and may not have `objectId`).